TST: ensure ``np.equal.reduce`` raises a ``TypeError``

Saransh-cpp opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Closes #20929 

Added a couple of tests to ensure that `np.equal.reduce` raises `TypeError`, and the developers are notified if the API changes.

I am not very sure if I should be creating a new method for this test. Please let me know if I can add the tests in any existing method.

Thanks!

Saransh-cpp committed:
TST: ensure `np.equal.reduce` raises a `TypeError`

charris commented:
close/reopen

charris commented on numpy/core/tests/test_umath.py:
```diff
@@ -267,6 +267,14 @@ def __ne__(self, other):
+        # without specifying the dtype
+        a = np.array([0, 0])
+        assert_equal(np.equal.reduce(a, dtype=bool), True)
+        assert_raises(TypeError, np.equal.reduce, a)
```
This is causing a segfault in the debug builds.

seberg replied:
hmmm sounds like hitting an assert, probably just a wrong one...

Saransh-cpp replied:
Thank you for the review, @charris and @seberg! I am very new to the C++ and Python combination. Could you please help me with the segfault, or link a similar error and fix so that I can have look? I tried running the testing suite by removing the `assert_raises` test and it does pass.

seberg replied:
Hmmmpf, should be able to get a traceback using gdb: https://numpy.org/devdocs/dev/development_environment.html#debugging

But I cannot reproduce with a python3.9 debug version and see nothing with valgrind either.  Can you undo the last commit, because I don't know anymore what is running on CI.  At least if it is a linux debug print, we can also hook in gdb in CI (although I have to look up how to every time).
If you can reproduce a crash with `-g` locally, then running `gdb` will tell us more.

seberg replied:
Oh, I can reproduce, it requires a full test-suite run though... maybe something wrong related to the dispatching caching.

seberg replied:
OK, I think the assert is indeed incorrect and it should be something like:
```diff
diff --git a/numpy/core/src/umath/dispatching.c b/numpy/core/src/umath/dispatching.c
index 5aecdd1fc..f2c6deada 100644
--- a/numpy/core/src/umath/dispatching.c
+++ b/numpy/core/src/umath/dispatching.c
@@ -1016,8 +1016,13 @@ promote_and_get_ufuncimpl(PyUFuncObject *ufunc,
             signature[i] = (PyArray_DTypeMeta *)PyTuple_GET_ITEM(all_dtypes, i);
             Py_INCREF(signature[i]);
         }
-        else {
-            assert((PyObject *)signature[i] == PyTuple_GET_ITEM(all_dtypes, i));
+        else if ((PyObject *)signature[i] != PyTuple_GET_ITEM(all_dtypes, i)) {
+            /*
+             * If signature is forced we may find an incompatible loop that
+             * would be picked by promotion (signature not forced), reject it.
+             */
+            raise_no_loop_found_error(ufunc, (PyObject **)op_dtypes);
+            return NULL;
         }
     }
```
We still get a slightly different error depending on the code path, but that is because the "normal" error is less precise.

seberg replied:
OK, maybe not the ideal thing, but I pushed a second commit that applies that fix (with a slightly different comment) and adds a more explicit test.

Saransh-cpp replied:
Thank you for fixing this, @seberg!! I was caught up with college and couldn't undo the commit at that time (thanks for fixing that too!)

seberg committed:
BUG: Replace assert with correct error

If we cache a promoted version of the loop, that promoted can mismatch
the correct one.  This ends up being rejected later in the legacy paths
(should be the only path currently used), but we should reject it here
(or in principle we could reject it after cache lookup, but we are fixing
up the operation DTypes here anyway, so we are looking at the signature).

A call sequence reproducing this directly is:

np.add(1, 2, signature=(bool, int, None))  # should fail

np.add(True, 2)  # A promoted loop
np.add(1, 2, signature=(bool, int, None))  # should still fail

Not that the errors differ, because the first one comes from the old
type resolution code and is currently less precise
